src/include/ipc/dai.h

```diff
@@ -61,30 +61,42 @@ enum sof_ipc_dai_type {
 	SOF_DAI_INTEL_ALH,		/**< Intel ALH */
 	SOF_DAI_IMX_SAI,                /**< i.MX SAI */
 	SOF_DAI_IMX_ESAI,               /**< i.MX ESAI */
+	SOF_DAI_MULTIDAI = 0xFFFF       /**< config with multiple elements of the same dai type */
 };
 
 /* general purpose DAI configuration */
 struct sof_ipc_dai_config {
 	struct sof_ipc_cmd_hdr hdr;
-	uint32_t type;		/**< DAI type - enum sof_ipc_dai_type */
+
+	/* FIXME endianness */
+	uint16_t type;		/**< DAI type - enum sof_ipc_dai_type */
+	uint16_t subdai_type;	/**< only used in multi-dai cases */
 	uint32_t dai_index;	/**< index of this type dai */
 
 	/* physical protocol and clocking */
 	uint16_t format;	/**< SOF_DAI_FMT_ */
-	uint16_t reserved16;	/**< alignment */
+
+	/* we can either have multiple sub-DAIs or multiple configs, not both */
+	uint16_t num_payloads; 	/**< number of configs or number of sub-DAIs */
 
 	/* reserved for future use */
 	uint32_t reserved[8];
 
 	/* HW specific data */
 	union {
-		struct sof_ipc_dai_ssp_params ssp;
-		struct sof_ipc_dai_dmic_params dmic;
-		struct sof_ipc_dai_hda_params hda;
-		struct sof_ipc_dai_alh_params alh;
-		struct sof_ipc_dai_esai_params esai;
-		struct sof_ipc_dai_sai_params sai;
+		struct sof_ipc_dai_ssp_params ssp[0];
+		struct sof_ipc_dai_dmic_params dmic[0];
+		struct sof_ipc_dai_hda_params hda[0];
+		struct sof_ipc_dai_alh_params alh[0];
+		struct sof_ipc_dai_esai_params esai[0];
+		struct sof_ipc_dai_sai_params sai[0];
```

Comment:
@plbossart when there are multiple configs per elementary DAI, the topology also defines the default config to be chosen. Should we add that info to this struct as well?

Reply:
thinking more about it, if we do add such a field, maybe we can use the DAI_CONFIG IPC to change the chosen config from the driver.

Reply:
and while I'm at it, can I also request we use the C99 syntax for flex arrays and leave out the 0? I know we use the C90 syntax pretty much everywhere else, but maybe its time to switch. But not a big deal

Reply:
somehow compilers don't like the [] case, I get this:

error: flexible array member in union
   struct sof_ipc_dai_ssp_params ssp[];

I am starting to wonder is this the right syntax, or if I should use something like

````c
	union {
		struct sof_ipc_dai_ssp_params ssp;
		struct sof_ipc_dai_dmic_params dmic;
		struct sof_ipc_dai_hda_params hda;
		struct sof_ipc_dai_alh_params alh;
		struct sof_ipc_dai_esai_params esai;
		struct sof_ipc_dai_sai_params sai;
	} u[0];
````